Convert autodownloads from master switch to per-podcast setting

newhinton opened the pull request:
### Description


### Checklist
<!-- 
  To help us keep the issue tracker clean and work as efficient as possible,
  please make sure that you have done all of the following.
  You can tick the boxes below by placing an x inside the brackets like this: [x]
-->
- [x] I have read the contribution guidelines: https://github.com/AntennaPod/AntennaPod/blob/develop/CONTRIBUTING.md#submit-a-pull-request
- [ ] I have performed a self-review of my code
- [ ] I have run the automated code checks using `./gradlew checkstyle spotbugsPlayDebug spotbugsDebug :app:lintPlayDebug`
- [ ] My code follows the style guidelines of the AntennaPod project: https://github.com/AntennaPod/AntennaPod/wiki/Code-style
- [x] I have mentioned the corresponding issue and the relevant keyword (e.g., "Closes: #xy") in the description (see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] If it is a core feature, I have added automated tests


## General Outline
This is the first draft of the discussed changes to the handling of the settings around auto-downloads.

It removes the "global" setting that enables or disables autodownloads. Autodownloads are now always enabled.
The general behaviour did not change however, since the new "default"-setting is still off. All new podcasts will follow this new setting. Only if changed, the behaviour changes. 

This change allows any podcast to enable the download feature, regardless of the global setting.


## Migrations
For this feature, there had to be a database upgrade. Each value of the 'KEY_AUTO_DOWNLOAD_ENABLED'-table column has been migrated from a 1 to 2. This is because the default value of that column is 1, and this needs to be reflected in the model. (alternatively the column could be dropped, migrated and recreated with the proper default value of 2, but that seemed drastic)

The result is that all old podcasts will be "manually enabled or disabled" after this. They will not follow the global default anymore.

### Example Podcast Settings:
<img src="https://github.com/user-attachments/assets/6aa88b10-a329-483f-a4e3-e3eac8361289" height="360vh">

### User Implications
1. When autodownloads were disabled, formerly enabled podcasts will now be downloaded again
2. "Untouched" podcasts that have their settings never altered, will not follow this new default.
There might be more. Those need to be discussed.


## Todo
There is still some code-cleanup todo, and some handling is missing. There is also a few places where i need help with my general architecture decisions, but overall, for a first review this works well enough. (I also need to work on the above checkboxes)

ByteHamster commented on storage/preferences/src/main/java/de/danoeh/antennapod/storage/preferences/UserPreferences.java:
```diff
@@ -530,10 +532,15 @@ public static int getEpisodeCacheSize() {
         return Integer.parseInt(prefs.getString(PREF_EPISODE_CACHE_SIZE, "20"));
     }
 
+    @Deprecated
```
I would remove this instead of deprecating. You can use PreferenceUpgrader to migrate to the new preference key. Then future developers cannot accidentally use the old value

newhinton replied:
Yeah, that is still a leftover. I wanted to get a mvp out before i started fixing tests and all those occurences.

newhinton commented on storage/database/src/main/java/de/danoeh/antennapod/storage/database/DBUpgrader.java:
```diff
@@ -349,6 +349,9 @@ static void upgrade(final SQLiteDatabase db, final int oldVersion, final int new
             db.execSQL("ALTER TABLE " + PodDBAdapter.TABLE_NAME_FEED_ITEMS
                     + " ADD COLUMN " + PodDBAdapter.KEY_PODCASTINDEX_TRANSCRIPT_TYPE + " TEXT");
         }
+        if (oldVersion < 3060000) {
```
If i remember correctly, we need to store three states on the feed item itself.
0=DL Disabled
1=Follow global
2=DL Enabled

If we don't differentiate between 1 and 2, it will not be possible for single feeds to be "undecided". They would be either enabled, or disabled. Though i'll check this again

ByteHamster replied:
Sorry, what I meant was: the 3 states are fine, but I don't think the database values of existing users need to be migrated. For most users they already are "1", so they would default to the global default. No need to manually migrate them to "2" (overwrite global with "on")

newhinton replied:
Ah! Sure, that's possible.

My initial idea was that since we don't know the users intent, we just force enable them again after the migration, this way they will always be downloaded. We might download too much, but not too few.

Your suggestion works too, but since the new default is "off", we don't download anything, not even formerly enabled entries. However, if we migrate the "autodownload enabled" variable to the new one, that will work out to be exactly the same set of downloads than before. 

That sounds better to me than my implementation. What do you think?

ByteHamster replied:
Yes, that's what I meant. Keep the behavior for existing users the same. I think we could even use the existing preference, no need to create a new one with the name "global"

newhinton commented:
@ByteHamster It seems the failing test does not actually show the error log. Do you know if this is related to my code?

ByteHamster commented:
The failing test does show the error log, you just need to scroll up a little:

> error: invalid method reference .until(UserPreferences::isEnableAutodownload);

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/drawer/NavListAdapter.java:
```diff
@@ -216,7 +216,7 @@ private void bindNavView(@StringRes int title, int position, NavHolder holder) {
                 holder.count.setVisibility(View.VISIBLE);
             }
-        } else if (tag.equals(CompletedDownloadsFragment.TAG) && UserPreferences.isEnableAutodownload()) {
+        } else if (tag.equals(CompletedDownloadsFragment.TAG)) {
```
Hmm, so that icon would be shown for people who don't use auto-download at all. We need to come up with a better way to indicate this. (also, people don't seem to understand this icon anyway, so maybe it needs to be changed completely)

newhinton replied:
Yes. I have not touched the limitation-part at all, so this shows up when that limit is reached. But this would be expected anyway, so i don't think that is a problem besides the general issue with it.

ByteHamster replied:
I think it might be better to remove the icon then. Otherwise it could be confusing to users who do not use auto-download.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/feed/preferences/FeedSettingsFragment.java:
```diff
@@ -462,25 +462,21 @@ private void setupKeepUpdatedPreference() {
-
         private void setupAutoDownloadPreference() {
-            SwitchPreferenceCompat pref = findPreference("autoDownload");
+            SwitchPreferenceCompat pref = findPreference(PREF_AUTODOWNLOAD);
```
The preference should be a 3-state choice: Yes, No, Global. Similar to the auto-delete setting on the same screen.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/preferences/AutomaticDeletionPreferencesFragment.java:
```diff
@@ -34,10 +34,11 @@ private void checkItemVisibility(boolean autoDeleteEnabled) {
-            findPreference(UserPreferences.PREF_EPISODE_CLEANUP).setEnabled(false);
-            findPreference(UserPreferences.PREF_EPISODE_CLEANUP).setSummary(R.string.auto_download_disabled_globally);
-        }
+        // todo: handle autodownload ui changes here
```
todo :)

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/Feed.java:
```diff
@@ -176,10 +176,11 @@ public Feed(String url, String lastModified, String title) {
     /**
      * This constructor is used for requesting a feed download (it must not be used for anything else!). It should be
      * used if the title of the feed is already known.
+     * Todo: check if the change to DEFAULT is correct
```
I would say the change is correct:

> for requesting a feed download (it must not be used for anything else!)

So it shouldn't really matter, but I think default is always a good default :)

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedPreferences.java:
```diff
@@ -78,10 +78,33 @@ public static SkipSilence fromCode(int code) {
+    public enum AutoDownloadSetting {
+        DISABLED(0),
+        ENABLED(2),
+        // the reason for this is that the database column uses 1 as the default value.
+        // migrating the cells to "2" is more easy than changing the default value.
```
This comment is useful for reviewing, but does not need to be kept in the source code forever. Please remove for the final version

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedPreferences.java:
```diff
@@ -78,10 +78,33 @@ public static SkipSilence fromCode(int code) {
+        ENABLED(2),
+        // the reason for this is that the database column uses 1 as the default value.
+        // migrating the cells to "2" is more easy than changing the default value.
+        DEFAULT(1);
```
Everywhere else (speed, auto delete, episode action, etc), we call this "global", not "default".

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedPreferences.java:
```diff
@@ -168,12 +191,30 @@ public void setFeedID(long feedID) {
-        return autoDownload;
+    public boolean getAutoDownload(boolean globalDefault) {
+        switch (this.autoDownload) {
+            case ENABLED -> {
+                return true;
+            }
+            case DISABLED -> {
+                return false;
+            }
+            default -> {
+                return globalDefault;
+            }
```
Hmm, this needs more lines than the "normal" switch syntax. Not sure if we should do this.

newhinton replied:
What do you mean with "normal" switch syntax? Do you mean two if's and then a return similar to line 290 in that file? Also, does it matter how "long" this switch is, if it is easily understandable?

ByteHamster replied:
Originally, switches in Java looked like this:
```
switch(...) {
    case ENABLED:
        return false;
```

Java 12 introduced switch expressions:

```
return switch(...) {
    case ENABLED -> false;
```

What is used here is some kind of mix between a switch expression (with the arrow `->`) and the statement (with colon and return). Everywhere else in the app, we use the "original" switch syntax.

newhinton replied:
Oh yeah. I am switching so often between java and kotlin, i just use what the ide suggests :D

I updated it to switch expressions because it fits good here, but i can change it back to the original one if you like.

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedPreferences.java:
```diff
@@ -168,12 +191,30 @@ public void setFeedID(long feedID) {
 
     public void setAutoDownload(boolean autoDownload) {
-        this.autoDownload = autoDownload;
+        if (autoDownload) {
+            this.autoDownload = AutoDownloadSetting.ENABLED;
+        } else  {
+            this.autoDownload = AutoDownloadSetting.DISABLED;
+        }
```
There should be a way to set it back to global, so this method should take an `AutoDownloadSetting`

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/FeedPreferences.java:
```diff
@@ -168,12 +191,30 @@ public void setFeedID(long feedID) {
+        }
+    }
+
+    public AutoDownloadSetting getAutoDownloadRaw() {
```
For other settings, we basically only have the raw method (not called raw) and handle the global default outside this class

newhinton replied:
I would like to keep this (but maybe rename it). My reasoning is that often i don't really care what the global value is, just the end-result. (I would have moved the `UserPreferences.defaultAutodownloadState()` also inside this method, but that would add a dependency to that module)

This makes the abstraction a bit easier since i don't need to reimplement the calculation everywhere.

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/SubscriptionsFilter.java:
```diff
@@ -73,10 +73,12 @@ public List<Feed> filter(List<Feed> items, Map<Long, Integer> feedCounters) {
         for (Feed item : items) {
             FeedPreferences itemPreferences = item.getPreferences();
 
+            // todo: UserPreferences
```
todo

ByteHamster replied:
Hmm, so the reason for these dependency warnings is to avoid developers adding logic to the model objects. A few years ago, the model objects did all kinds of things such as network and database calls. See https://github.com/AntennaPod/AntennaPod/issues/4661

The only place where this filter method is used is DbReader. Maybe you could move the actual filtering logic to a new class that is located in the same module as DbReader? I think that might be a better location anyway, to reduce the amount of logic in the data model. (for episodes, we do the filtering in sql already, which might be the better option anyway. Having the method closer to DbReader could make this change easier in the future)

ByteHamster commented on net/download/service/src/main/java/de/danoeh/antennapod/net/download/service/episode/autodownload/EpisodeCleanupAlgorithmFactory.java:
```diff
@@ -4,9 +4,6 @@
 
 public abstract class EpisodeCleanupAlgorithmFactory {
     public static EpisodeCleanupAlgorithm build() {
-        if (!UserPreferences.isEnableAutodownload()) {
-            return new APNullCleanupAlgorithm();
-        }
```
We need to be careful with the default here. It looks like this means that AntennaPod will start deleting episodes after the upgrade, even if you don't use auto-download. If auto-download is disabled globally, we should change the default to `UserPreferences.EPISODE_CLEANUP_NULL` when upgrading (`PreferenceUpgrader`) and make sure that the default for new installs is also `UserPreferences.EPISODE_CLEANUP_NULL`

newhinton replied:
As far as i understand the cleanup code, the default for new installations should already be "null".

I have added the upgrades, but this probably needs to be tested to really make sure it does what you intend it to do.

ByteHamster commented on storage/database/src/main/java/de/danoeh/antennapod/storage/database/PodDBAdapter.java:
```diff
@@ -471,7 +471,7 @@ public void setFeedPreferences(FeedPreferences prefs) {
         }
         ContentValues values = new ContentValues();
-        values.put(KEY_AUTO_DOWNLOAD_ENABLED, prefs.getAutoDownload());
+        values.put(KEY_AUTO_DOWNLOAD_ENABLED, prefs.getAutoDownloadRaw().state);
```
Everywhere else we call the integer version "code", not "state"

ByteHamster commented on storage/preferences/src/main/java/de/danoeh/antennapod/storage/preferences/UserPreferences.java:
```diff
@@ -98,7 +98,7 @@ public abstract class UserPreferences {
     public static final String PREF_EPISODE_CLEANUP = "prefEpisodeCleanup";
     public static final String PREF_EPISODE_CACHE_SIZE = "prefEpisodeCacheSize";
-    public static final String PREF_ENABLE_AUTODL = "prefEnableAutoDl";
+    public static final String PREF_AUTODL_DEFAULT = "prefEnableAutoDl";
```
Please call it global, not default, for consistency with other settings

ByteHamster commented on ui/preferences/src/main/res/xml/preferences_autodownload.xml:
```diff
@@ -1,13 +1,7 @@
-            android:title="@string/pref_automatic_download_title"
-            search:summary="@string/pref_automatic_download_sum"
-            android:defaultValue="false"/>
+    <!-- Removed Preference: prefEnableAutoDl -->
```
No need to leave a tombstone comment :) The setting is just gone :)

ByteHamster commented on ui/preferences/src/main/res/xml/preferences_autodownload.xml:
```diff
@@ -1,13 +1,7 @@
-        xmlns:search="http://schemas.android.com/apk/com.bytehamster.lib.preferencesearch">
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
 
-    <de.danoeh.antennapod.ui.preferences.preference.MasterSwitchPreference
```
This was the only use of `MasterSwitchPreference`. I think the class can therefore be removed.

ByteHamster commented on ui/preferences/src/main/res/xml/preferences_autodownload.xml:
```diff
@@ -20,4 +14,9 @@
             android:title="@string/pref_automatic_download_on_battery_title"
             android:summary="@string/pref_automatic_download_on_battery_sum"
             android:defaultValue="true"/>
+    <SwitchPreferenceCompat
```
I think it would fit better as the first entry rather than the last

newhinton commented:
@ByteHamster You have mentioned in some commits that we need to be able to set the setting back to the global default. 

Is that actually the case, ux wise? In what instances would that be helpful, and does that justify the additional complexity? (UX-wise, not codewise)

I'd argue that once the setting has been changed, it's decoupled and requires a manual interaction again. I'd use it to remove an unwanted podcast from my subscriptions, but i am not sure when setting it back to global would be useful.

 @keunes what do you think about this, design wise?

ByteHamster commented:
I think @keunes and me discussed this some time ago. It would be frustrating for users to be able to change a setting to something but not being able to change it back.

A use-case could be to turn on/off all auto-downloads at once using the global setting while on vacation. In the current implementation, as soon as you have touched that setting once, you won't be able to do this anymore.

newhinton commented:
> A use-case could be to turn on/off all auto-downloads at once using the global setting while on vacation. In the current implementation, as soon as you have touched that setting once, you won't be able to do this anymore.

That does not seem to be useful, because to achieve that the user needs to reset every single changed podcast, before they go on vacation. if they even know that they did that. (It would be more useful to change the refresh-toggle to 'never')


I would suggest to keep it like it is at the moment, and see if users actually have a need for this setting. And if so, maybe a better way to do this is how dolphin(emulator) deals with this, where a long press resets to the default.

My reasoning is that it reduces the amount of ui-options (and by not having dialogs everywhere also the code complexity)

If you really think it should be a dialog i am going to buikd one, but i'd advise against it.

ByteHamster commented:
> That does not seem to be useful, because to achieve that the user needs to reset every single changed podcast, before they go on vacation.

No, they just need to keep them on "global". However, as soon as they touch the setting with your PR, this is no longer possible.

> and by not having dialogs everywhere also the code complexity

Using an explicit 3-state switch reduces the code complexity. We use that in many other settings already. Having one setting that works differently is inconsistent and confusing

newhinton commented:
> Having one setting that works differently is inconsistent and confusing

That is true. 

> No, they just need to keep them on "global". However, as soon as they touch the setting with your PR, this is no longer possible.

That is what i mean. As soon as one toggle has been changed, you can't control the podcasts in a "away for vacation"-way with the default state, because not all podcasts follow that. For that to work, the user has to change each podcast back, and undo that change after the vacation. What i am trying to say is: The default value would not be the proper place to archive that goal. That would be a different setting like the podcast refresh. Do you really think this is a feature users would actually use and warrant adding more options to the ui?

ByteHamster commented:
> Do you really think this is a feature users would actually use and warrant adding more options to the ui?

Yes, I'm sure. Not being able to change a setting back to what it was before is frustrating.

ByteHamster commented:
We have to watch out that this doesn't accidentally enable auto-delete for users who had the global "master switch" for auto-download (and therefore delete as well) disabled

keunes commented:
Sorry for the late follow-up here.

> I'd argue that once the setting has been changed, it's decoupled and requires a manual interaction again. I'd use it to remove an unwanted podcast from my subscriptions, but i am not sure when setting it back to global would be useful.
> 
> @keunes what do you think about this, design wise?

I agree with @ByteHamster that a three-state setting would be useful, even if a slightly more complicated UX. The three-state is a common UX pattern in the app already, and provides more transparency about the current setting and flexibility for changing behaviour (e.g. when starting to use AntennaPod and experimenting with the settings).

Thanks for your work on this @newhinton!

newhinton commented:
@ByteHamster 

> We have to watch out that this doesn't accidentally enable auto-delete for users who had the global "master switch" for auto-download (and therefore delete as well) disabled

Which setting exactly do you mean?
- Delete after Playing
- Auto delete from local folders
- Delete before auto download

The last one is handled by one of your comments in the code itself, and its corresponding change/update for the preference.

newhinton commented:
@ByteHamster I have implemented some changes you mentioned, and fixed some issues. There are still some comments open where i need help in making a decision, but those should be minor.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/feed/preferences/FeedSettingsFragment.java:
```diff
@@ -462,42 +462,54 @@ private void setupKeepUpdatedPreference() {
+                    case "global":
+                        feedPreferences.setAutoDownload(FeedPreferences.AutoDownloadSetting.GLOBAL);
+                        break;
+                    case "always":
```
always/never is a bit unexpected here. I would just call it enabled/disabled, like the setting is already called in Java

ByteHamster commented:
Thanks! Will be released in AntennaPod 3.7.0. This was a larger and more complex change than I expected. I think this will be one of the top changes highlighted in the changelog of the next release :)
